codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/endpoint/EndpointsDecorator.kt

```diff
@@ -49,120 +62,128 @@ class EndpointsDecorator : RustCodegenDecorator<ClientProtocolGenerator, ClientC
+                withInlineModule(EndpointTests) {
+                    generator.testGenerator()(this)
+                }
             }
-
-            else -> emptySection
         }
     }
 
-    private fun builderFields(params: Parameters, section: OperationSection.MutateInput) = writable {
-        val memberParams = idx.getContextParams(operationShape)
-        val builtInParams = params.toList().filter { it.isBuiltIn }
-        // first load builtins and their defaults
-        builtInParams.forEach { param ->
-            val defaultProviders = rulesEngineBuiltInResolvers.mapNotNull { it.defaultFor(param, section.config) }
-            if (defaultProviders.size > 1) {
-                error("Multiple providers provided a value for the builtin $param")
-            }
-            defaultProviders.firstOrNull()?.also { defaultValue ->
-                rust(".set_${param.name.rustName()}(#W)", defaultValue)
+    /**
+     * Creates an `<crate>::endpoint_resolver::Params` structure in make operation generator. This combines state from the
+     * client, the operation, and the model to create parameters.
+     *
+     * Example generated code:
+     * ```rust
+     * let _endpoint_params = crate::endpoint_resolver::Params::builder()
+     *     .set_region(Some("test-region"))
+     *     .set_disable_everything(Some(true))
+     *     .set_bucket(input.bucket.as_ref())
+     *     .build();
+     * ```
+     */
+    class InjectEndpointInMakeOperation(
+        private val ctx: ClientCodegenContext,
+        private val typesGenerator: EndpointTypesGenerator,
+        private val operationShape: OperationShape,
+    ) :
+        OperationCustomization() {
+
+        private val idx = ContextIndex.of(ctx.model)
+
+        override fun section(section: OperationSection): Writable {
+            val codegenScope = arrayOf(
+                "Params" to typesGenerator.paramsStruct(),
+                "BuildError" to ctx.runtimeConfig.operationBuildError(),
+            )
+            return when (section) {
+                is OperationSection.MutateInput -> writable {
+                    rustTemplate(
+                        """
+                        let endpoint_params = #{Params}::builder()#{builderFields:W}.build()
+                            .map_err(#{BuildError}::other)?;
+                        let endpoint_result = ${section.config}.endpoint_resolver.resolve_endpoint(&endpoint_params);
+                        """,
+                        "builderFields" to builderFields(typesGenerator.params, section),
+                        *codegenScope,
+                    )
+                }
+
+                is OperationSection.MutateRequest -> writable {
+                    // insert the endpoint resolution _result_ into the bag (note that this won't bail if endpoint
+                    // resolution failed)
+                    rustTemplate("${section.request}.properties_mut().insert(endpoint_params);")
```

Comment:
hmm, why? (it also enables logging / debugging). Because the params would be stack allocated otherwise? We don't end up cloning them just making one allocation and putting them in the bag

Reply:
It looked like there was a `.clone()` happening for every `String` parameter after placing the params in the property bag.